src/System.Linq/src/System/Linq/Take.SpeedOpt.cs

```diff
@@ -12,5 +13,58 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(count > 0);
+
+            if (source is IPartition<TSource> partition)
+            {
+                int length = partition.GetCount(true);
+                if (length > 0)
+                {
+                    return length - count > 0 ? partition.Skip(length - count) : partition;
+                }
+                else if (length == 0)
+                {
+                    return partition;
+                }
+            }
+            else if (source is IList<TSource> sourceList)
+            {
+                if (sourceList.Count > count)
+                {
+                    return new ListPartition<TSource>(sourceList, sourceList.Count - count, sourceList.Count);
```

Comment:
sourceList.Count is an interface call, and this makes that call three times.  It should instead be called once and stored into a local.

Reply:
Right